### PR TITLE
Implementing rolling min/max functions that can retain the original type

### PR DESCRIPTION
**Pull request by joshuastorck**
- [ ] Enhances previous bugfix to #12373 
- [ ] Added test_rolling_min_max_test_types to test_windows
- [ ] passes `git diff upstream/master | flake8 --diff`
- [ ] Added documentation to computation.rst for the new as_float argument and updated whatsnew for v0.18.0
  - Changed the rolling min/max functions in algos.pyx so
    that they use a cython fused type as input instead of a float64
    so that the function can accept arrays of any numeric type
  - Merged the functionality of rolling min/max into a common
    function with branches based on whether or not it's running
    min/max
  - When running rolling min/max for intergral types and there
    are not enough minimum periods, the output values returned
    are zero
  - Added a unit test to test_moments to make sure that rolling
    min/max works for all integral types and float32/64
  - Updated computations and whatsnew doc

### CODE REVIEW

**Comment by kawochen** [comment]
> When running rolling min/max for intergral types and there are not enough minimum periods, the output values returned are zero

I don't think 0 is sensible, but I'm not sure how else it can be done either without integral nulls.

**Comment by kawochen** [comment]
> Merged the functionality of rolling min/max into a common function with branches based on whether or not it's running min/max

the branching can be decided very soon (your `is_max`), so you don't need to branch in the loop

**Comment by joshuastorck** [comment]
If you branched earlier, you would effectively be passing around a function pointer. That doesn't give the compiler any opportunity to do something smarter, like a conditional move, which in this case would be pretty easy to generate. 

Ideally, much of this code should be written in native C++ with templates using C++11 lambdas. That would make it possible to eliminate branching completely at compile time.

File: doc/source/computation.rst [hunk 1]
```diff
@@ -250,6 +250,10 @@ accept the following arguments:
   result is NA)
 - ``center``: boolean, whether to set the labels at the center (default is False)
 
+.. note::
+
+   The ``min`` and ``max`` functions will by default return the result as a float. For integer inputs, integer outputs can be obtained by passing True as the ``as_float`` argument.
```
Comment (jreback) [inline doc/source/computation.rst]: take this out, we won't be having an ``as_float`` argument

File: doc/source/whatsnew/v0.18.0.txt [hunk 2]
```diff
@@ -1004,6 +1004,7 @@ Other API Changes
 - ``.describe()`` will now properly handle bool dtype as a categorical (:issue:`6625`)
 - More helpful error message with an invalid ``.transform`` with user defined input (:issue:`10165`)
 - Exponentially weighted functions now allow specifying alpha directly (:issue:`10789`) and raise ``ValueError`` if parameters violate ``0 < alpha <= 1`` (:issue:`12492`)
+- ``Rolling.min`` and ``Rolling.max`` now take an as_float argument, that when False, will cause those functions to return output that has the same value as the input type. This is
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: move this all to 0.18.1 (and take out this entry)

File: doc/source/whatsnew/v0.18.0.txt [hunk 3]
```diff
@@ -1290,3 +1291,4 @@ Bug Fixes
 - Bug when specifying a UTC ``DatetimeIndex`` by setting ``utc=True`` in ``.to_datetime`` (:issue:`11934`)
 - Bug when increasing the buffer size of CSV reader in ``read_csv`` (:issue:`12494`)
 - Bug when setting columns of a ``DataFrame`` with duplicate column names (:issue:`12344`)
+- Bug in ``.rolling.min`` and ``.rolling.max`` where passing columns of type float32 raised a Value error (:issue:`12373`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: this one is ok

File: pandas/core/window.py [hunk 4]
```diff
@@ -148,16 +148,17 @@ def _shallow_copy(self, obj=None, **kwargs):
                 kwargs[attr] = getattr(self, attr)
         return self._constructor(obj, **kwargs)
 
-    def _prep_values(self, values=None, kill_inf=True, how=None):
+    def _prep_values(self, values=None, kill_inf=True, how=None,
+                     as_float=True):
```
Comment (jreback) [inline pandas/core/window.py]: remove the ``as_float`` arg (it will simply coerce ATM for integer types, no easy way around this)

**Comment by kawochen** [comment]
@joshuastorck  Ah! I see!

**Comment by joshuastorck** [comment]
Any additional work needed on this one?

**Comment by jreback** [comment]
will take a look - can u rebase in master

**Comment by joshuastorck** [comment]
Just rebased and pushed

**Comment by jreback** [comment]
also pls ``git rebase -i master`` then push upstream (with ``-f``).

**Comment by jreback** [comment]
nvm. you are only 1 or 2 commits behind.

File: pandas/algos.pyx [hunk 5]
```diff
@@ -1625,29 +1625,38 @@ def roll_median_c(ndarray[float64_t] arg, int win, int minp):
...
-
 from libc cimport stdlib
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
-def roll_max(ndarray[float64_t] a, int window, int minp):
-    "Moving max of 1d array of dtype=float64 along axis=0 ignoring NaNs."
-    cdef np.float64_t ai, aold
+def roll_max(ndarray[numeric] a, int window, int minp):
+    "Moving max of 1d array of any numeric type along axis=0 ignoring NaNs."
```
Comment (jreback) [inline pandas/algos.pyx]: bonus if we can document the params (even though its an internal function)

File: pandas/algos.pyx [hunk 6]
```diff
@@ -1659,146 +1668,85 @@ def roll_max(ndarray[float64_t] a, int window, int minp):
...
     minp = _check_minp(window, minp, n0)
     with nogil:
-        ring = <pairs*>stdlib.malloc(window * sizeof(pairs))
+        ring = <numeric*>stdlib.malloc(window * sizeof(numeric))
+        death = <Py_ssize_t*>stdlib.malloc(window * sizeof(Py_ssize_t))
         end = ring + window
         last = ring
 
-        minpair = ring
+        minvalue = ring
```
Comment (jreback) [inline pandas/algos.pyx]: minvalue does not seem to be declared

Comment (joshuastorck) [inline pandas/algos.pyx]: cython declares it automatically

**Comment by jreback** [comment]
I think you need to take out some of the ``_ensure_float64`` that were implemented in the related PR.

I would expect this to return ``float32`` no?

```
In [4]: Series([1,2,3,4],dtype='float32').rolling(window=2).max()
Out[4]: 
0    NaN
1    2.0
2    3.0
3    4.0
dtype: float64

```

pretty much everything else is going to be upcast though

File: pandas/algos.pyx (hunk 6 above)
Comment (jreback) [inline pandas/algos.pyx]: hmm, but you declare everything else, any problem with declaring it?

Comment (joshuastorck) [inline pandas/algos.pyx]: nope, I'll put it in and add the docs.

**Comment by joshuastorck** [comment]
That was why I originally had an as_float argument, which was in there for backwards compatibility. Are you saying that you are ok with breaking type compatibility?

**Comment by jreback** [comment]
@joshuastorck not sure what you mean. We are *now* breaking type compat by casing to ``float64``. if you can preserve the type then by all means do it. You don't need a flag for that it should just work. Of course for integers ATM this won't work, but that's a different issue.

**Comment by joshuastorck** [comment]
So you want this code in window.py:

```
        # GH #12373 : rolling functions error on float32 data
        # make sure the data is coerced to float64
        if com.is_float_dtype(values.dtype):
            values = com._ensure_float64(values)
        elif com.is_integer_dtype(values.dtype):
            values = com._ensure_float64(values)

```

To change to this?

```
        # GH #12373 : rolling functions error on float32 data
        # make sure the data is coerced to float64
        if com.is_integer_dtype(values.dtype):
            values = com._ensure_float64(values)

```

**Comment by joshuastorck** [comment]
Also,I only changed the roll_min/max functions. Wouldn't that code change break all of the other rolling functions?

**Comment by jreback** [comment]
hmm, yes ideally that would be the change. But you are right the rest have not been converted. ok then.

pls squash / ping when green.

**Comment by jreback** [comment]
ok, I linked the master issue: https://github.com/pydata/pandas/issues/8659

so next up is fixing the remainder of the rolling functions, so we don't need to upcast floats. :)

**Comment by joshuastorck** [comment]
Rebased and Travis CI is green

File: pandas/algos.pyx [hunk 7]
```diff
@@ -1747,58 +1686,85 @@ def roll_min(np.ndarray[np.float64_t, ndim=1] a, int window, int minp):
...
+            if numeric in cython.floating:
+                y[i0] = NaN
+            else:
+                y[i0] = 0
 
         stdlib.free(ring)
+        stdlib.free(death)
     return y
 
+cdef double_t _get_max(object skiplist, int nobs, int minp):
```
Comment (jreback) [inline pandas/algos.pyx]: did you add this? I don't see this (or ``_get_min``) used AT ALL in the codebase. any idea?

Comment (jreback) [inline pandas/algos.pyx]: you don't need to fix, just lmk.

Comment (joshuastorck) [inline pandas/algos.pyx]: I think it was there before and was removed but I accidentally added it in a merge. I won’t have a chance to get to this today. Can you accept the pull request and I’ll make a separate PR to remove it?

Comment (jreback) [inline pandas/algos.pyx]: I am going to merge this and take it out. thxnks

**Comment by jreback** [comment]
@joshuastorck thanks for the fixes!

#8659 (using fused types in the rest of the algos) is waiting for you!

**Comment by jreback** [comment]
@joshuastorck  going to start going thru your changes for support on [this](https://github.com/jreback/pandas/commit/5985ae7803801264240809faf4a90f02c27836e5)
